OfflineAudioContext/explainer.md

```diff
@@ -0,0 +1,203 @@
+# WebAudio OfflineAudioContext.startRendering() streaming output
+
+## Authors:
+
+- [Matt Birman](mailto:mattbirman@microsoft.com)
+
+## Table of Contents
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Introduction
+
+WebAudio `OfflineAudioContext.startRendering()` allocates an `AudioBuffer` large enough to hold the entire render WebAudio graph before returning. For example, a 4 hour audio graph at 48 kHz with 4 channels will create gigabytes of in-memory float32 data in the `AudioBuffer`. This behaviour makes the API is unsuitable for very long offline renders or very large channel/length combinations. There is no simple way to chunk the output or consume it as a stream.
+
+The [spec](https://webaudio.github.io/web-audio-api/#dom-offlineaudiocontext-startrendering) explicitly states at step 5: "Create a new AudioBuffer ... with ... length and sampleRate ... Assign this buffer to an internal slot" which means the API design currently mandates the full buffer allocation.
+
+The participants on the [GitHub discussion](https://github.com/WebAudio/web-audio-api/issues/2445) agree that incremental delivery of data is necessary. Either streaming chunks of rendered audio or dispatching data in bits rather than everything at once so that memory usage is bounded and the data can be processed/consumed as it is produced.
+
+## User-Facing Problem
+
+The user in this context is the web developer using the WebAudio API to perform media processing workflows. Ideally developers could use the feature-rich WebAudio API for realtime and faster-than-realtime processing, without taking a dependency on a 3rd party library. However, in reality, the current WebAudio OfflineAudioContext API is not suitable for faster-than-raltime processing so the developer needs to create a WASM audio processing library or use an existing 3rd party dependency to achieve this goal.
+
+### Goals
+
+- Allow streaming data out of a WebAudio `OfflineAudioContext.startRendering()` for rendering large WebAudio graphs faster-than-realtime
+
+### Non-goals
+
+- Change the existing `startRendering()` behavior, this API change is additive
+
+## Proposed Approach
+
+The preferred approach is to allow `startRendering()` to be configured to stream output via a `StartRenderingOptions` object which will have a `mode` property. This `mode` can be set to `"stream"` which will stream data rather than allocating an `AudioBuffer` up front.
+
+The object allows for future flexibility of the new API surface.
+
+```typescript
+interface StartRenderingOptions {
+    mode: "audiobuffer" | "stream"
+}
+
+interface OfflineAudioContext {
+  startRendering(options?: startRenderingOptions}
```

Comment:
It feels like the discussion on https://github.com/WebAudio/web-audio-api/issues/2445 was leaning towards having a new set of dedicated methods for streaming, like:
```js
partial interface OfflineAudioContext {
  ReadableStream startRenderingStream(optional OfflineAudioRenderingStreamOptions);  

  // If startRenderingStream() was called, this closes the stream and the context will be closed.
  Promise<void> stopRenderingStream();
};
```

Should we suggest this instead as our first proposal?